### PR TITLE
Domain Name Mapping for windows login

### PR DESCRIPTION
**Pull request by Deepesh125**
### Description

In SQL Server, User can provide partial domain name/ Netbios domain name in all DDL related to WINDOWS LOGIN (CREATE/ALTER/DROP LOGIN). SQL Server can still figure out the full domain name from the AD. For example, BABEL\tom, BABEL.COM\tom, adtest.BABEL.COM\tom all are valid in SQL Server.

It is not possible to translate this netbios domain name to fully qualified domain name without knowing the LDAP credential of AD setup. So, In Babelfish (and in PG), the engine doesn’t have the ability to automatically figure out the full domain name from partial domain name.

To resolve, We need create domain mapping catalog which can maintain this Netbios domain name to FQDN. This commit introduces new catalog sys.babelfish_domain_mapping to hold the Netbios domain name to FQDN mapping. As well as introduces procedures, sys.babelfish_add_domain_mapping_entry to add domain mapping entry, sys.babelfish_remove_domain_mapping_entry to remove domain mapping entry and  sys.babelfish_reset_domain_mapping to reset the domain mapping entries.

With this change, when convertToUPN() is called to convert the user format, convertToUPN will internally call get_fully_qualified_domain_name() to retrieve FQDN from customer provided Netbios domain name.

For example, If user has added [babel] to [babel.internal] mapping and If customer uses [babel\test] login then it will internally converted to test@BABEL.INTERNAL. 

### Issues Resolved

Task: BABEL-3847
Sub-task: BABEL-3863

Signed-off-by: Dipesh Dhameliya <dddhamel@amazon.com>

### Test Scenarios Covered ###
* **Use case based -**

Added test cases to verify function of domain mapping as well as CREATE\DROP login.

* **Boundary conditions -**

* **Arbitrary inputs -**

* **Negative test cases -**

* **Minor version upgrade tests -**

* **Major version upgrade tests -**

* **Performance tests -**

* **Tooling impact -**

* **Client tests -**

### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: contrib/babelfishpg_tsql/src/rolecmds.c [hunk 1]
```diff
@@ -1672,3 +1695,315 @@ is_active_login(Oid role_oid)
...
+			ereport(ERROR,
+					(errcode(ERRCODE_INSUFFICIENT_PRIVILEGE),
+					 errmsg("Current login %s does not have permission to add new domain mapping entry",
+					 GetUserNameFromId(GetSessionUserId(), true))));
+
+	bbf_domain_mapping_rel = table_open(get_bbf_domain_mapping_oid(), RowExclusiveLock);
+
+	/* Write catalog entry */
+	new_record = palloc0(sizeof(Datum) * BBF_DOMAIN_MAPPING_NUM_COLS);
+	new_record_nulls = palloc0(sizeof(bool) * BBF_DOMAIN_MAPPING_NUM_COLS);
```
Comment (KushaalShroff) [inline contrib/babelfishpg_tsql/src/rolecmds.c:2043]: Please free them

**Review by KushaalShroff — CHANGES_REQUESTED** [review]

File: contrib/babelfishpg_tsql/src/rolecmds.c (hunk 1 above)
Comment (Deepesh125) [reply contrib/babelfishpg_tsql/src/rolecmds.c:2043]: Done

File: contrib/babelfishpg_tsql/src/rolecmds.c [hunk 2]
```diff
@@ -1858,6 +1931,230 @@ get_roleform_ext(char *login)
...
+
+	MemSet(new_record_nulls, false, sizeof(new_record_nulls));
+
+	new_record[0] = PG_GETARG_DATUM(0);
+	new_record[1] = PG_GETARG_DATUM(1);
+	validateNetBIOS(TextDatumGetCString(new_record[0]));
+	validateFQDN(TextDatumGetCString(new_record[1]));
```
Comment (Deepesh125) [inline contrib/babelfishpg_tsql/src/rolecmds.c]: Let's move this checks to line no. 2028, basically before opening table.

File: contrib/babelfishpg_tsql/src/rolecmds.h [hunk 3]
```diff
@@ -24,6 +24,11 @@
 #define LOGON_NAME_MAX_LEN 21
 #define LOGON_NAME_MIN_LEN 0
 
+#define NETBIOS_NAME_MAX_LEN 15
+#define NETBIOS_NAME_MIN_LEN 0
```
Comment (Deepesh125) [inline contrib/babelfishpg_tsql/src/rolecmds.h]: should be 1

File: contrib/babelfishpg_tsql/src/rolecmds.c [hunk 4]
```diff
@@ -1858,6 +1931,230 @@ get_roleform_ext(char *login)
...
+validateNetBIOS(char* netbios)
+{
+	int len = strlen(netbios);
+	int i = 0;
+
+	if (len >= NETBIOS_NAME_MAX_LEN || len <= NETBIOS_NAME_MIN_LEN)
+		ereport(ERROR,
+			(errcode(ERRCODE_INVALID_NAME),
+				errmsg("The NetBIOS name '%s' has invalid length. NetBIOS name length should be between %d and %d.",
+				netbios, (NETBIOS_NAME_MIN_LEN + 1), (NETBIOS_NAME_MAX_LEN - 1))));
```
Comment (Deepesh125) [inline contrib/babelfishpg_tsql/src/rolecmds.c]: use NETBIOS_NAME_MIN_LEN and NETBIOS_NAME_MAX_LEN. Why do we need to NETBIOS_NAME_MAX_LEN - 1?

File: contrib/babelfishpg_tsql/src/rolecmds.h [hunk 5]
```diff
@@ -24,6 +24,11 @@
...
+
+#define FQDN_NAME_MAX_LEN 128
+#define FQDN_NAME_MIN_LEN 1
```
Comment (Deepesh125) [inline contrib/babelfishpg_tsql/src/rolecmds.h]: should be 2

File: contrib/babelfishpg_tsql/src/rolecmds.c [hunk 6]
```diff
@@ -1858,6 +1931,230 @@ get_roleform_ext(char *login)
...
+validateFQDN(char* fqdn)
+{
+	int len = strlen(fqdn);
+	int i = 1;
+
+	if (len >= FQDN_NAME_MAX_LEN || len <= FQDN_NAME_MIN_LEN)
+		ereport(ERROR,
+			(errcode(ERRCODE_INVALID_NAME),
+				errmsg("The FQDN '%s' has invalid length. FQDN length should be between %d and %d.",
+				fqdn, (FQDN_NAME_MIN_LEN + 1), (FQDN_NAME_MAX_LEN - 1))));
```
Comment (Deepesh125) [inline contrib/babelfishpg_tsql/src/rolecmds.c]: Use FQDN_NAME_MIN_LEN and FQDN_NAME_MAX_LEN

File: contrib/babelfishpg_tsql/src/rolecmds.c [hunk 7]
```diff
@@ -1858,6 +1931,230 @@ get_roleform_ext(char *login)
...
+
+	if (!((fqdn[0] >= 65 && fqdn[0] <= 90) || (fqdn[0] >= 97 && fqdn[0] <= 122) || (fqdn[0] >= 48 && fqdn[0] <= 57)))
```
Comment (Deepesh125) [inline contrib/babelfishpg_tsql/src/rolecmds.c]: replace all this with equivalent letin char, for exampple, fqdn[0] >= 'a'

File: contrib/babelfishpg_tsql/src/rolecmds.c [hunk 8]
```diff
@@ -1858,6 +1931,230 @@ get_roleform_ext(char *login)
...
+		ereport(ERROR, (errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+			errmsg("'%s' is not a valid FQDN. It must start with alphabetical or numeric character .", fqdn)));
```
Comment (Deepesh125) [inline contrib/babelfishpg_tsql/src/rolecmds.c]: unnecessary space before full stop at the end.

File: contrib/babelfishpg_tsql/src/rolecmds.c (hunk 2 above)
Comment (ahmed-shameem) [reply contrib/babelfishpg_tsql/src/rolecmds.c]: Done!

File: contrib/babelfishpg_tsql/src/rolecmds.h (hunk 3 above)
Comment (ahmed-shameem) [reply contrib/babelfishpg_tsql/src/rolecmds.h]: Done!

File: contrib/babelfishpg_tsql/src/rolecmds.c (hunk 4 above)
Comment (ahmed-shameem) [reply contrib/babelfishpg_tsql/src/rolecmds.c]: Done!

File: contrib/babelfishpg_tsql/src/rolecmds.h (hunk 5 above)
Comment (ahmed-shameem) [reply contrib/babelfishpg_tsql/src/rolecmds.h]: Done!

File: contrib/babelfishpg_tsql/src/rolecmds.c (hunk 6 above)
Comment (ahmed-shameem) [reply contrib/babelfishpg_tsql/src/rolecmds.c]: Done!

File: contrib/babelfishpg_tsql/src/rolecmds.c (hunk 7 above)
Comment (ahmed-shameem) [reply contrib/babelfishpg_tsql/src/rolecmds.c]: Done!

File: contrib/babelfishpg_tsql/src/rolecmds.c (hunk 8 above)
Comment (ahmed-shameem) [reply contrib/babelfishpg_tsql/src/rolecmds.c]: Done!

File: test/JDBC/upgrade/13_6/schedule [hunk 9]
```diff
@@ -301,4 +301,4 @@ sys-sql_modules_before-14_7-or-15_2
 sys-all_sql_modules_before-14_5
 BABEL_OBJECT_ID
 test_windows_login_before_15_2
-BABEL_OBJECT_NAME
\ No newline at end of file
+BABEL_OBJECT_NAME
```
Comment (macwoody) [inline test/JDBC/upgrade/13_6/schedule]: why do you have theses changes, new and old look the same to me

**Review by macwoody — COMMENTED** [review]
only one comment, plz help check it.

File: test/JDBC/upgrade/13_6/schedule (hunk 9 above)
Comment (Deepesh125) [reply test/JDBC/upgrade/13_6/schedule]: This must have happened while resolving merge conflicts. It should not cause any issue.

File: contrib/babelfishpg_tsql/src/pl_handler.c [hunk 10]
```diff
@@ -2315,10 +2315,10 @@ static void bbf_ProcessUtility(PlannedStmt *pstmt,
 							stmt->role = convertToUPN(orig_loginname);
 
 							/*
-							* Check for duplicate login
-							*/
+							 * Check for duplicate login
+							 */
 							if (get_role_oid(stmt->role, true) != InvalidOid)
-						  		ereport(ERROR, (errcode(ERRCODE_DUPLICATE_OBJECT), 
+								ereport(ERROR, (errcode(ERRCODE_DUPLICATE_OBJECT),
```
Comment (kuntalghosh) [inline contrib/babelfishpg_tsql/src/pl_handler.c:2321]: Unnecessary change.

File: contrib/babelfishpg_tsql/src/rolecmds.c [hunk 11]
```diff
@@ -1858,6 +1933,240 @@ get_roleform_ext(char *login)
...
+
+	if (len > NETBIOS_NAME_MAX_LEN || len < NETBIOS_NAME_MIN_LEN)
+		ereport(ERROR,
+			(errcode(ERRCODE_INVALID_NAME),
+				errmsg("The NetBIOS name '%s' has invalid length. NetBIOS name length should be between %d and %d.",
+				netbios, NETBIOS_NAME_MIN_LEN, NETBIOS_NAME_MAX_LEN)));
+
+	if (netbios[0] == '.')
+		ereport(ERROR, (errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+			errmsg("'%s' is not a valid NetBIOS name. It must not start with '.' .", netbios)));
```
Comment (kuntalghosh) [inline contrib/babelfishpg_tsql/src/rolecmds.c:1954]: Why is this scenario separately tested from the conditions mentioned in the while loop next?

**Review by kuntalghosh — COMMENTED** [review]
Do we have test cases for each of the error messages that we've introduced?

File: contrib/babelfishpg_tsql/src/rolecmds.c (hunk 11 above)
Comment (ahmed-shameem) [reply contrib/babelfishpg_tsql/src/rolecmds.c:1954]: We could test that in the while loop itself, but as netbios name can not start with '.', we are checking it separately. Or else, in the loop we would have been checking whether  i == 0 or not every single time. With this check, we won't be doing that anymore.

File: contrib/babelfishpg_tsql/src/pl_handler.c (hunk 10 above)
Comment (ahmed-shameem) [reply contrib/babelfishpg_tsql/src/pl_handler.c:2321]: Maybe it came up during resolving merge conflicts.

**Review by kuntalghosh — APPROVED** [review]

File: contrib/babelfishpg_tsql/src/pl_handler.c (hunk 10 above)
Comment (Deepesh125) [reply contrib/babelfishpg_tsql/src/pl_handler.c:2321]: This change is there to fix indentation.

**Review by kuntalghosh — APPROVED** [review]
